Enhancement to crankcase heater

yujiex opened the pull request:
Pull request overview
---------------------
Crankcase heaters use a simple electrical circuit to create heat within the crankcase when the compressor of the chiller or heat pump is not in use. The heat keeps the refrigerant from condensing and prevents the refrigerant from migrating through the seals into the oil. Poor control of the crankcase heater could result in excessive energy consumption. Some of the existing products start to provide variable heater power based on different outdoor temperature conditions. Currently, EnergyPlus can only model a constant power crankcase heater with a temperature cutoff. This feature proposes to enhance the crankcase heater in EnergyPlus to allow for variable heater power, which enables to model realistic performance of heat pumps.

- Fixes #10137

# testing with a testfile

Running HeatPump_variableCrankcaseHeaterCapacity.idf, with a crankcase heater capacity curve of y = 60 - 2x. The following plot shows the relationship between outdoor temperature and heating coil crankcase heater electricity rate, limited to the time period with runtime fraction = 0.

<img width="829" alt="image" src="https://user-images.githubusercontent.com/13421161/235204708-30677e65-b02b-4057-8664-cb89bc32d8d7.png">

The following is the simulation result of HeatPumpVSAS_variableCrankcaseHeaterCapacity.idf with variable speed coils using the same crankcase heater capacity curve. Column C computes the crankcase heater power before adjusting for the runtime fraction

<img width="951" alt="image" src="https://user-images.githubusercontent.com/13421161/235273996-68c39f9d-5e9a-49eb-a364-c36c8488df54.png">

# Comparing output of a HeatPumpVSAS.idf and HeatPumpVSAS_variableCrankcaseHeaterCapacity.idf with constant curve

```
!- crankcase heater capacity in HeatPumpVSAS.idf 
  Coil:Heating:DX:VariableSpeed,
    Heat Pump DX Heating Coil 1,  !- Name
    Heating Coil Air Inlet Node,  !- Indoor Air Inlet Node Name
    SuppHeating Coil Air Inlet Node,  !- Indoor Air Outlet Node Name
    10.0,                    !- Number of Speeds {dimensionless}
    10.0,                    !- Nominal Speed Level {dimensionless}
    35000,                   !- Rated Heating Capacity At Selected Nominal Speed Level {W}
    1.7,                     !- Rated Air Flow Rate At Selected Nominal Speed Level {m3/s}
    HPACCOOLPLFFPLR,         !- Energy Part Load Fraction Curve Name
    ,                        !- Defrost Energy Input Ratio Function of Temperature Curve Name
    -5.0,                    !- Minimum Outdoor Dry-Bulb Temperature for Compressor Operation {C}
    ,                        !- Outdoor Dry-Bulb Temperature to Turn On Compressor {C}
    5.0,                     !- Maximum Outdoor Dry-Bulb Temperature for Defrost Operation {C}
    200.0,                   !- Crankcase Heater Capacity {W}


!- curve in HeatPumpVSAS_variableCrankcaseHeaterCapacity.idf
  Curve:Linear,
    heaterCapCurve,          !- Name
    200.0,                    !- Coefficient1 Constant
    0.0,                     !- Coefficient2 x
    -80,                     !- Minimum Value of x
    80;                      !- Maximum Value of x
```
As anticipated, the crankcase heater power are the same for the two cases
![no_curve_vs_const_200_curve](https://github.com/NREL/EnergyPlus/assets/13421161/dff81fc1-e909-45ab-b13b-a6da2e594725)

Comparing a curve with a constant capacity of 200 and a curve with a constant capacity of 100

![const_200_vs_const_100](https://github.com/NREL/EnergyPlus/assets/13421161/42798224-9587-41cd-9964-f3af155a388f)
 
# Summary of some identified issues 

## 1. The crankcase heater for the variable speed heat pump water heater might be off all the time

In the file VariableSpeedCoils.cc, the following chunk of code (in the function CalcVarSpeedHPWH) here modifies the temporary variable `CrankcaseHeatingPower` which is then used in the calculation of `state.dataVariableSpeedCoils->VarSpeedCoil(DXCoilNum).CrankcaseHeaterPower` later on. It seems that the check of whether the temperature is appropriate for the crankcase heater to operate is only executed when EvapInletNode is 0. But this variable gets value from `state.dataVariableSpeedCoils->VarSpeedCoil(DXCoilNum).AirInletNodeNum` which is from a required field. So it can’t be 0 in normal cases. This might mean the crankcase heater power will always be 0
```
        // check if indoor evaporator or outdoor evaporator
        CrankcaseHeatingPower = 0.0;
        if (EvapInletNode != 0) {
            state.dataVariableSpeedCoils->LoadSideInletDBTemp = state.dataLoopNodes->Node(EvapInletNode).Temp;
            state.dataVariableSpeedCoils->LoadSideInletHumRat = state.dataLoopNodes->Node(EvapInletNode).HumRat;
            LoadPressure = state.dataLoopNodes->Node(EvapInletNode).Press;
            // prevent the air pressure not given
            if (LoadPressure < 10.0) LoadPressure = state.dataEnvrn->OutBaroPress;

            state.dataVariableSpeedCoils->LoadSideInletWBTemp = state.dataLoopNodes->Node(EvapInletNode).OutAirWetBulb;
            state.dataVariableSpeedCoils->LoadSideInletEnth = state.dataLoopNodes->Node(EvapInletNode).Enthalpy;
        } else {
            state.dataVariableSpeedCoils->LoadSideInletDBTemp = state.dataEnvrn->OutDryBulbTemp;
            state.dataVariableSpeedCoils->LoadSideInletHumRat = state.dataEnvrn->OutHumRat;
            LoadPressure = state.dataEnvrn->OutBaroPress;
            state.dataVariableSpeedCoils->LoadSideInletWBTemp = state.dataEnvrn->OutWetBulbTemp;
            state.dataVariableSpeedCoils->LoadSideInletEnth = state.dataEnvrn->OutEnthalpy;

            // Initialize crankcase heater, operates below OAT defined in input deck for HP DX heating coil
            if (state.dataEnvrn->OutDryBulbTemp < state.dataVariableSpeedCoils->VarSpeedCoil(DXCoilNum).MaxOATCrankcaseHeater) {
                CrankcaseHeatingPower = state.dataVariableSpeedCoils->VarSpeedCoil(DXCoilNum).CrankcaseHeaterCapacity;
                if (state.dataVariableSpeedCoils->VarSpeedCoil(DXCoilNum).CrankcaseHeaterCapacityCurveIndex > 0) {
                    CrankcaseHeatingPower = Curve::CurveValue(state,
                                                              state.dataVariableSpeedCoils->VarSpeedCoil(DXCoilNum).CrankcaseHeaterCapacityCurveIndex,
                                                              state.dataEnvrn->OutDryBulbTemp);
                }
            };
        } 
```

## 2. Some inconsistency in what temperature should be compared to the maximum temperature threshold below which the crankcase heater should operate

See this issue: https://github.com/NREL/EnergyPlus/issues/10052

**NOTE: ENHANCEMENTS MUST FOLLOW A SUBMISSION PROCESS INCLUDING A FEATURE PROPOSAL AND DESIGN DOCUMENT PRIOR TO SUBMITTING CODE**

### Pull Request Author
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [x] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
 - [x] Label the PR with at least one of: Defect, Refactoring, NewFeature, Performance, and/or DoNoPublish
 - [ ] Pull requests that impact EnergyPlus code must also include unit tests to cover enhancement or defect repair
 - [ ] Author should provide a "walkthrough" of relevant code changes using a GitHub code review comment process
 - [ ] If any diffs are expected, author must demonstrate they are justified using plots and descriptions
 - [ ] If changes fix a defect, the fix should be demonstrated in plots and descriptions
 - [ ] If any defect files are updated to a more recent version, upload new versions here or on DevSupport
 - [ ] If IDD requires transition, transition source, rules, ExpandObjects, and IDFs must be updated, and add IDDChange label
 - [ ] If structural output changes, add to output rules file and add OutputChange label
 - [ ] If adding/removing any LaTeX docs or figures, update that document's CMakeLists file dependencies

### Reviewer
This will not be exhaustively relevant to every PR.
 - [ ] Perform a Code Review on GitHub
 - [ ] If branch is behind develop, merge develop and build locally to check for side effects of the merge
 - [ ] If defect, verify by running develop branch and reproducing defect, then running PR and reproducing fix
 - [ ] If feature, test running new feature, try creative ways to break it
 - [ ] CI status: all green or justified
 - [ ] Check that performance is not impacted (CI Linux results include performance check)
 - [ ] Run Unit Test(s) locally
 - [ ] Check any new function arguments for performance impacts
 - [ ] Verify IDF naming conventions and styles, memos and notes and defaults
 - [ ] If new idf included, locally check the err file and other outputs

yujiex committed:
initial commit

yujiex committed:
correct typo

yujiex committed:
Merge remote-tracking branch 'origin/develop' into crankcaseHeater

yujiex committed:
Initial commmit of test file idf

yujiex committed:
idd change, input reading and test for Coil:Cooling/Heating:DX

yujiex committed:
Fix unit test

yujiex committed:
idd change, input reading and testing of Coil:WaterHeating:AWHP

yujiex committed:
clang-format

yujiex committed:
add testfile to CMakeLists.txt

yujiex committed:
Apply crankcase heater capacity curve to DXCoils

yujiex committed:
Apply crankcase heater curve to variable speed coils

yujiex committed:
Initial commit of tesetfile variableSpeedCoil crankcaseheater curve

mjwitte commented on idd/Energy+.idd.in:
```diff
@@ -51383,6 +51383,10 @@ Coil:Cooling:DX:CurveFit:Performance,
        \note is used as Reheat mode.
        \type object-list
        \object-list DXCoolingOperatingModeNames
+   A8; \field Outdoor Temperature Dependent Crankcase Heater Capacity Curve Name
```
To be consistent with other curve names this should be
"Crankcase Heater Capacity Function of Temperature Curve Name"
ref. 
Coil:Cooling:DX:SingleSpeed, Total Cooling Capacity Function of Temperature Curve Name
Coil:Cooling:DX:CurveFit:Speed, Total Cooling Capacity Modifier Function of Temperature Curve Name

Also this should be tagged with 
```
       \type object-list
       \object-list UnivariateFunctions
```

yujiex replied:
Right, I'll change the name to "Crankcase Heater Capacity Function of Temperature Curve Name" to match the existing patterns and add the object-list tag.

mjwitte replied:
> Right, I'll change the name to "Crankcase Heater Capacity Function of Temperature Curve Name" to match the existing patterns and add the object-list tag.

These changes are still needed. Also, for each affected object, \min-fields should be increased by one (unless the new field is already beyond the current min-fields value).

yujiex replied:
right right, I'll change min-fields too for ones too for those with the new curve field not beyond the current min-field.

mjwitte commented on idd/Energy+.idd.in:
```diff
@@ -53562,12 +53574,16 @@ Coil:Cooling:DX:VariableSpeed,
         \note quadratic curve = a + b*ffa + c*ffa**2
         \note cubic curve = a + b*ffa + c*ffa**2 + d*ffa**3
         \note ffa = Fraction of the full load Air Flow
+   A50; \field Outdoor Temperature Dependent Crankcase Heater Capacity Curve Name
```
Hmm, the group of speed fields is currently not extensible, so technically, this is ok to add at the end of the object, but it just seems lost here. @Myoldmopar do we care?

yujiex replied:
Yeah, I'm not certain whether I should put it at the end or near the other crankcase heater fields.

mjwitte commented on src/EnergyPlus/Coils/CoilCoolingDXCurveFitPerformance.hh:
```diff
@@ -117,6 +118,7 @@ struct CoilCoolingDXCurveFitPerformance
     std::string name;
     Real64 crankcaseHeaterCap = 0.0;
     Real64 crankcaseHeaterPower = 0.0;
+    Real64 crankcaseHeaterCapacityCurveIndex = 0.0;
```
Integer

yujiex replied:
It was fixed in another commit. I'll push it soon. I was trying to figure out some issues with the variable speed coils curve application.

Myoldmopar replied:
> It was fixed in another commit.

This is still not an integer.  Did you by chance forget to push a commit?

mjwitte commented on tst/EnergyPlus/unit/Coils/CoilCoolingDXCurveFitOperatingMode.unit.cc:
```diff
@@ -209,3 +210,63 @@ TEST_F(CoilCoolingDXTest, CoilCoolingDXCurveFitOperatingMode_Sizing)
+    EXPECT_EQ("COIL COOLING DX 1", thisCoil.name);
+    EXPECT_EQ("COIL COOLING DX CURVE FIT PERFORMANCE 1", thisCoil.performance.name);
+    EXPECT_EQ("HEATERCAPCURVE", Curve::GetCurveName(*state, thisCoil.performance.crankcaseHeaterCapacityCurveIndex));
+}
```
Add a return to the last line.

yujiex replied:
an explicit return like this `return;`? @mjwitte

mjwitte replied:
Sorry, no, a <cr> return. Look at the display in the PR and you'll see a red circle at the end, because the last line of the file is not terminated fully.

yujiex replied:
Oh! I understand now. I'll add a new line in the end.

mjwitte commented on tst/EnergyPlus/unit/DXCoils.unit.cc:
```diff
@@ -1075,6 +1076,860 @@ TEST_F(EnergyPlusFixture, DXCoilEvapCondPumpSizingTest)
+        ",",
+        ",",
+        ",",
+        ",",
+        ",",
+        ",",
+        ",",
+        ",",
+        ",",
+        ",",
+        ",",
+        "heaterCapCurve5;                          !- Outdoor Temperature Dependent Crankcase Heater Capacity Curve Name",
```
And this shows why the new field should be placed before the speed groups for variable speed and multi-speed coils, even if it forces a transition rule. @Myoldmopar ?

yujiex replied:
yeah, this is kind of tedious.

mjwitte commented on src/EnergyPlus/Coils/CoilCoolingDXCurveFitPerformance.cc:
```diff
@@ -124,6 +124,10 @@ void CoilCoolingDXCurveFitPerformance::instantiateFromInputSpec(EnergyPlus::Ener
         setOperMode(state, this->alternateMode2, 3);
     }
 
+    if (!input_data.outdoor_temperature_dependent_crankcase_heater_capacity_curve_name.empty()) {
+        this->crankcaseHeaterCapacityCurveIndex =
+            Curve::GetCurveIndex(state, input_data.outdoor_temperature_dependent_crankcase_heater_capacity_curve_name);
+    }
```
The input processing needs more steps here (and everywhere) to check if a valid curve was found and to check the curve dimensions. For example, see [here](https://github.com/NREL/EnergyPlus/blob/557201b3c0ec9d8b5e7235cd4c8d893c4e01fef4/src/EnergyPlus/DXCoils.cc#L1320-L1335). For this application, any uni-variate curve or table is acceptable.

yujiex replied:
good point. I'll add in the curve validations.

mjwitte reviewed (commented):
@yujiex A few comments below. Also, I don't see any doc changes yet.

yujiex committed:
Modify testfiles

yujiex committed:
add explicit return to unit test

yujiex committed:
Change curve field name to match existing pattern

yujiex committed:
add curve validation

yujiex committed:
Merge remote-tracking branch 'origin/develop' into crankcaseHeater

nrel-bot-2 commented:
@yujiex it has been 7 days since this pull request was last updated.

nrel-bot-2b commented:
@yujiex it has been 7 days since this pull request was last updated.

nrel-bot-2c commented:
@yujiex it has been 7 days since this pull request was last updated.

nrel-bot-2 commented:
@yujiex it has been 7 days since this pull request was last updated.

yujiex committed:
Use the same temperature var for the curve and threshold

yujiex committed:
Add end of file new line

yujiex committed:
Add unit tests for applying the heater capacity curve

yujiex committed:
Merge remote-tracking branch 'origin/develop' into crankcaseHeater

yujiex committed:
update doc

yujiex committed:
fix idf and tests from idd change in merged develop

yujiex committed:
clang-format

yujiex commented:
Hi @mjwitte and @Myoldmopar, I've added some unit tests and updated some documentation. Would you mind taking another look at this PR. Thanks :)

yujiex committed:
add comments to unit test

yujiex committed:
Add heater curve code in Coil:Cooling:DX:CurveFit:Performance

rraustad commented on src/EnergyPlus/Coils/CoilCoolingDXCurveFitPerformance.cc:
```diff
@@ -351,6 +366,9 @@ void CoilCoolingDXCurveFitPerformance::simulate(EnergyPlus::EnergyPlusData &stat
         this->crankcaseHeaterPower = this->crankcaseHeaterCap;
+        if (this->crankcaseHeaterCapacityCurveIndex > 0) {
+            this->crankcaseHeaterPower = Curve::CurveValue(state, this->crankcaseHeaterCapacityCurveIndex, state.dataEnvrn->OutDryBulbTemp);
+        }
```
This should be `*=` to multiply by the curve value.

rraustad replied:
Oh, I see, this is not a normalized curve. It probably should be so that it would be multiplied by the crankcase heater capacity.

    Curve:Linear,
      heaterCapCurve,          !- Name
      60.0,                    !- Coefficient1 Constant
      -2.,                     !- Coefficient2 x
      -30,                     !- Minimum Value of x
      40;                      !- Maximum Value of x

rraustad replied:
Use this or something like this:

    Curve:Linear,
      heaterCapCurve,
      0.571428571428572,
      -0.0142857142857143,
      -30,
      40;

![image](https://github.com/NREL/EnergyPlus/assets/5778156/427d3329-48fc-42aa-bd67-91c93c4eecb0)

yujiex replied:
I'll change it to a normalized curve and multiply that by the crankcase heater capacity. @rraustad

yujiex replied:
@rraustad Forgot to ping you. I have updated the curve to be multiplied by the crankcase heater capacity. Would you mind taking another look? Thanks :)

rraustad replied:
Yes, this looks better.

yujiex committed:
Merge remote-tracking branch 'origin/develop' into crankcaseHeater

yujiex committed:
Change to normalized curve

rraustad commented on testfiles/HeatPumpVSAS_variableCrankcaseHeaterCapacity.idf:
```diff
@@ -0,0 +1,1798 @@
+    ,                        !- Condenser Air Inlet Node Name
+    AirCooled,               !- Condenser Type
+    ,                        !- Evaporative Condenser Pump Rated Power Consumption {W}
+    ,                        !- Crankcase Heater Capacity {W}
```
For Coil:Cooling:DX:VariableSpeed, the crankcase heater as a function of temperature curve needs to be moved here. And a value for Crankcase Heater Capacity should be entered, otherwise CC heater power will be 0. A transition rule will be needed.

rraustad commented on idd/Energy+.idd.in:
```diff
@@ -53854,12 +53866,16 @@ Coil:Cooling:DX:VariableSpeed,
+   A50; \field Outdoor Temperature Dependent Crankcase Heater Capacity Curve Name
+        \note A Curve:* or Table:Lookup object encoding the relationship between
+        \note the crankcase heater capacity and the outdoor air temperature. When this field is
+        \note missing or empty, constant crankcase heater capacity will be assumed.
```
This input field should be moved up above the speed fields. It looks strange here. A transition to insert a field is not that hard. And once you do that, consider moving this field next to the Crankcase Heater Capacity input for all coil types. I won't force that, but it would then be consistent for all coil types.

yujiex replied:
So I guess all these "Outdoor Temperature Dependent Crankcase Heater Capacity Curve Name" should be moved to the front near the Crankcase Heater Capacity fields. That does make it easier to use.

yujiex replied:
@rraustad @mjwitte I have moved the "Outdoor Temperature Dependent Crankcase Heater Capacity Curve Name" to right after the "Crankcase Heater Capacity" field and added the transition code. Would you mind taking another look? Thanks. Currently, there are two regression diffs (regression.SingleFamilyHouse_HP_Slab and regression.SingleFamilyHouse_HP_Slab_Dehumidification) with idd difference due to the moving of the curve name field.

```
--- 
+++ 
@@ -3823,6 +3823,7 @@
     ,                        !- Outdoor Dry-Bulb Temperature to Turn On Compressor {C}
     5.0,                     !- Maximum Outdoor Dry-Bulb Temperature for Defrost Operation {C}
     200.0,                   !- Crankcase Heater Capacity {W}
+    ,                        !- Outdoor Temperature Dependent Crankcase Heater Capacity Curve Name
     10.0,                    !- Maximum Outdoor Dry-Bulb Temperature for Crankcase Heater Operation {C}
     ReverseCycle,            !- Defrost Strategy
     OnDemand;                !- Defrost Control
@@ -3842,6 +3843,7 @@
     Zone Exhaust Node1_unit1,!- Evaporator Air Inlet Node Name
     Heat Pump Water Heater Fan Inlet Node_unit1,  !- Evaporator Air Outlet Node Name
     ,                        !- Crankcase Heater Capacity {W}
+    ,                        !- Outdoor Temperature Dependent Crankcase Heater Capacity Curve Name
     10,                      !- Maximum Ambient Temperature for Crankcase Heater Operation {C}
     WetBulbTemperature,      !- Evaporator Air Temperature Type for Curve Objects
     HPWH-Htg-Cap-fT,         !- Heating Capacity Function of Temperature Curve Name
```

yujiex committed:
move curve to front, Coil:Cooling:DX:SingleSpeed

yujiex committed:
move curve to front, Coil:Cooling:DX:MultiSpeed

yujiex committed:
move coil to front, Coil:Heating:DX:SingleSpeed

yujiex committed:
adjust > NumAlphas threshold for Cooling:DX:SingleSpeed

yujiex committed:
adjust NumAlphas threshold in Coil:Cooling:DX:MultiSpeed

yujiex committed:
move curve to front, Coil:Heating:DX:MultiSpeed

yujiex committed:
move curve to front, Coil:Cooling:DX:TwoStageWithHumidityControlMode

yujiex committed:
move curve to front, Coil:WaterHeating:AirToWaterHeatPump:Pumped

yujiex committed:
move curve to front Coil:WaterHeating:AirToWaterHeatPump:Wrapped

yujiex committed:
move curve to front, Coil:Cooling:DX:VariableSpeed

yujiex committed:
move curve to front, Coil:Heating:DX:VariableSpeed

yujiex committed:
move curve to front, Coil:WaterHeating:AirToWaterHeatPump:VariableSpeed

yujiex committed:
move curve to front, Coil:Cooling:DX:CurveFit:Performance

yujiex committed:
fix testfile .idf with inserted field

yujiex committed:
fix unit test with added field

yujiex committed:
Merge remote-tracking branch 'origin/develop' into crankcaseHeater

yujiex committed:
fix WaterThermalTanks unit test

yujiex committed:
Merge remote-tracking branch 'origin/develop' into crankcaseHeater

yujiex committed:
add new field to the expand object epfilter.f90

yujiex committed:
try fixing long string

yujiex committed:
clang-format

yujiex committed:
fix idf raw string issue

yujiex committed:
clang-format

yujiex committed:
add transition code

yujiex committed:
Merge remote-tracking branch 'origin/develop' into crankcaseHeater

yujiex committed:
unify input field name

it might need to be different in the future, but that will be taken care of in a
separate branch

yujiex commented:
Hi @mjwitte and @rraustad, are there other items I should fix other than moving the curve field to up front near the "Crankcase Heater Capacity". Also this is my first time writing idd transition code (I spied on Rich's previous PR to see how it works), do you spot any issues? Thanks :)

mjwitte committed:
Merge remote-tracking branch 'remotes/origin/develop' into crankcaseHeater

mjwitte reviewed (commented):
I've merged in develop and resolved the conflict. I'll push that now.
Unit test run successfully.

Tested transition on each affected object type. Coil:Cooling:DX:VariableSpeed has a glitch. Example file PackagedTerminalHeatPumpVSAS.idf fails after transition.
```
   ** Severe  ** GetVarSpeedCoilInput: Coil:Cooling:DX:VariableSpeed="ZONE1PTHPDXCOOLCOIL", missing
   **   ~~~   ** ...required Speed 10 Energy Input Ratio Function of Air Flow Fraction Curve Name is blank.
   ** Severe  ** GetVarSpeedCoilInput: Coil:Cooling:DX:VariableSpeed="ZONE2PTHPDXCOOLCOIL", missing
   **   ~~~   ** ...required Speed 10 Energy Input Ratio Function of Air Flow Fraction Curve Name is blank.
   ** Severe  ** GetVarSpeedCoilInput: Coil:Cooling:DX:VariableSpeed="ZONE3PTHPDXCOOLCOIL", missing
   **   ~~~   ** ...required Speed 10 Energy Input Ratio Function of Air Flow Fraction Curve Name is blank.
   ** Severe  ** GetVarSpeedCoilInput: COIL:HEATING:DX:VARIABLESPEED="ZONE1PTHPDXHEATCOIL", missing
   **   ~~~   ** ...required Speed 10 Energy Input Ratio Function of Air Flow Fraction Curve Name is blank.
   ** Severe  ** GetVarSpeedCoilInput: COIL:HEATING:DX:VARIABLESPEED="ZONE2PTHPDXHEATCOIL", missing
   **   ~~~   ** ...required Speed 10 Energy Input Ratio Function of Air Flow Fraction Curve Name is blank.
   ** Severe  ** GetVarSpeedCoilInput: COIL:HEATING:DX:VARIABLESPEED="ZONE3PTHPDXHEATCOIL", missing
   **   ~~~   ** ...required Speed 10 Energy Input Ratio Function of Air Flow Fraction Curve Name is blank.
   **  Fatal  ** GetVarSpeedCoilInput: Errors found getting input. Program terminates.
```

yujiex commented:
> I've merged in develop and resolved the conflict. I'll push that now. Unit test run successfully.
> 
> Tested transition on each affected object type. Coil:Cooling:DX:VariableSpeed has a glitch. Example file PackagedTerminalHeatPumpVSAS.idf fails after transition.
> 
> ```
>    ** Severe  ** GetVarSpeedCoilInput: Coil:Cooling:DX:VariableSpeed="ZONE1PTHPDXCOOLCOIL", missing
>    **   ~~~   ** ...required Speed 10 Energy Input Ratio Function of Air Flow Fraction Curve Name is blank.
>    ** Severe  ** GetVarSpeedCoilInput: Coil:Cooling:DX:VariableSpeed="ZONE2PTHPDXCOOLCOIL", missing
>    **   ~~~   ** ...required Speed 10 Energy Input Ratio Function of Air Flow Fraction Curve Name is blank.
>    ** Severe  ** GetVarSpeedCoilInput: Coil:Cooling:DX:VariableSpeed="ZONE3PTHPDXCOOLCOIL", missing
>    **   ~~~   ** ...required Speed 10 Energy Input Ratio Function of Air Flow Fraction Curve Name is blank.
>    ** Severe  ** GetVarSpeedCoilInput: COIL:HEATING:DX:VARIABLESPEED="ZONE1PTHPDXHEATCOIL", missing
>    **   ~~~   ** ...required Speed 10 Energy Input Ratio Function of Air Flow Fraction Curve Name is blank.
>    ** Severe  ** GetVarSpeedCoilInput: COIL:HEATING:DX:VARIABLESPEED="ZONE2PTHPDXHEATCOIL", missing
>    **   ~~~   ** ...required Speed 10 Energy Input Ratio Function of Air Flow Fraction Curve Name is blank.
>    ** Severe  ** GetVarSpeedCoilInput: COIL:HEATING:DX:VARIABLESPEED="ZONE3PTHPDXHEATCOIL", missing
>    **   ~~~   ** ...required Speed 10 Energy Input Ratio Function of Air Flow Fraction Curve Name is blank.
>    **  Fatal  ** GetVarSpeedCoilInput: Errors found getting input. Program terminates.
> ```

Thanks @mjwitte, I will modify the field name to "Crankcase Heater Capacity Function of Temperature Curve Name", and fix the transition of variable speed coils.

yujiex committed:
field name change to Crankcase Heater Capacity Function of ...

yujiex committed:
idd increment min-field by 1 for new field within min-field

rraustad commented on idd/Energy+.idd.in:
```diff
@@ -52242,7 +52242,7 @@ Coil:Cooling:DX:MultiSpeed,
        \memo between speeds. Modeled as a single coil (multi-speed compressor or multiple
        \memo compressors with row split or intertwined coil).
-       \min-fields 95
+       \min-fields 96
```
96 min-fields for MultiSpeed coil? If this is a 2-spd coil do I really have to enter data for speed 3 and 4? A 2-spd coil ends at N34, plus A24, the previous alpha, gives 58 as min-fields for 2-spd. What am I missing?

yujiex replied:
Make sense. I'll limit it to the first two speeds. That would be 59 (Numeric 34 and alpha 25).

mjwitte replied:
Yes, 96 min-fields is correct in order to set all of the default fields for higher speeds. min-fields doesn't require that the object be that long, but if you do have, say a 4-speed coil, the defaults won't get used for any blank fields that are after min-fields. Looking at the getinput for this coil, it uses whatever gets returned from `getObjectItem` in the numbers array. But if the actual idf object goes beyond the blank fields (which it will for each speed group) then the defaults *do* get filled in by `getObjectItem`. Sooo, never mind. This is fine. (Tested this with a modified version of MultiSpeedHP_StagedThermostat.idf to prove this to myself).

yujiex committed:
fix Cooling:DX:VariableSpeed version transition

yujiex committed:
Revert "fix Cooling:DX:VariableSpeed version transition"

This reverts commit baed7f97ac393c309620984b3a4441e33d21dd2b.

yujiex committed:
fix variable speed coil transition increment CurArgs

yujiex committed:
Merge remote-tracking branch 'origin/develop' into crankcaseHeater

yujiex committed:
reduce min-fields Cooling:DX:MultiSpeed (limit to 2 spd)

yujiex committed:
Merge remote-tracking branch 'origin/develop' into crankcaseHeater

yujiex committed:
Fix unit test after merging

mjwitte committed:
Merge remote-tracking branch 'remotes/origin/develop' into crankcaseHeater

mjwitte committed:
Minor rules edit

mjwitte reviewed (approved):
@yujiex This looks good.
I re-tested transition for all coil types - success!
Ran all unit tests - success!
Merged in develop and added some edits to the input rules file.
@Myoldmopar @rraustad Speak up if you want to review one more time. This will be the first transition rule(s) to merge in for v23.2. Only for idf, no epJSON transition required because it just adds a new optional field.

rraustad commented on src/EnergyPlus/DXCoils.cc:
```diff
@@ -4623,27 +4705,27 @@ void GetDXCoils(EnergyPlusData &state)
                 } else {
                     ShowSevereError(state, format("{}{}=\"{}\", invalid", RoutineName, CurrentModuleObject, thisDXCoil.Name));
-                    ShowContinueError(state, format("...not found {}=\"{}\".", cAlphaFields(12 + (I - 1) * 6), Alphas(15 + (I - 1) * 6)));
+                    ShowContinueError(state, format("...not found {}=\"{}\".", cAlphaFields(13 + (I - 1) * 6), Alphas(13 + (I - 1) * 6)));
```
Was this a mistake in code and was corrected? I see an alpha index of 12 and 15 before, and now I see 13 and 13. If this fixes a mistake, which it looks like it does, then that should be added to the PR description.

yujiex replied:
It should probably be 13 as the error message was stating what the current value of this field is, so the cAlphaFields and Alphas should have the same index. The 12th (one before) and 14th (one after) fields also have similar formats. So I corrected it.

rraustad commented on src/EnergyPlus/DXCoils.cc:
```diff
@@ -9319,6 +9401,9 @@ void CalcDoe2DXCoil(EnergyPlusData &state,
         CrankcaseHeatingPower = thisDXCoil.CrankcaseHeaterCapacity;
+        if (thisDXCoil.CrankcaseHeaterCapacityCurveIndex > 0) {
+            CrankcaseHeatingPower *= Curve::CurveValue(state, thisDXCoil.CrankcaseHeaterCapacityCurveIndex, CompAmbTemp);
+        }
```
This code could have probably been added to ReportDXCoil, instead of each calc funtion. It's OK as is, just noting an alternate way to account for the curve value.

yujiex replied:
yeah, I do see a lot of places with this type of calculation.

rraustad commented on src/EnergyPlus/DXCoils.hh:
```diff
@@ -151,16 +151,17 @@ namespace DXCoils {
+        Real64 CrankcaseHeaterCapacity;        // total crankcase heater capacity [W]
+        Real64 CrankcaseHeaterPower;           // report variable for average crankcase heater power [W]
+        Real64 MaxOATCrankcaseHeater;          // maximum OAT for crankcase heater operation [C]
+        int CrankcaseHeaterCapacityCurveIndex; // Crankcase heater power-temperature curve or table index
```
I am not suggesting a change here, but when adding new struct varaiables, declare inline (i.e., = 0) instead of adding to the constructor. Better yet, delete the constructor and decalre everything inline.

yujiex replied:
make sense. Should probably gradually phase these out and use inline.

rraustad commented on src/EnergyPlus/VariableSpeedCoils.cc:
```diff
@@ -1163,13 +1176,13 @@ namespace VariableSpeedCoils {
                         state,
                         format("{}{}=\"{}\", invalid", RoutineName, CurrentModuleObject, state.dataVariableSpeedCoils->VarSpeedCoil(DXCoilNum).Name));
-                    ShowContinueError(state, format("...not found {}=\"{}\".", cAlphaFields(14), AlphArray(9)));
+                    ShowContinueError(state, format("...not found {}=\"{}\".", cAlphaFields(14), AlphArray(10)));
```
Why does this cAlphaFields index point to a different input than AlphArray?

yujiex replied:
I need to fix this

rraustad commented on src/EnergyPlus/VariableSpeedCoils.cc:
```diff
@@ -2705,29 +2742,29 @@ namespace VariableSpeedCoils {
             } else {
                 CurveVal = CurveValue(state, state.dataVariableSpeedCoils->VarSpeedCoil(DXCoilNum).PLFFPLR, 1.0);
-                if (CurveVal > 1.10 || CurveVal < 0.90) {
+                if (CurveVal > 1.11 || CurveVal < 0.90) {
```
Oops! This should be 1.10 (i.e., 10% tolerance).

yujiex replied:
oh no I probably did some global replace and didn't see this ... I'll fix it.

rraustad commented on src/EnergyPlus/VariableSpeedCoils.cc:
```diff
@@ -2705,29 +2742,29 @@ namespace VariableSpeedCoils {
                                             CurrentModuleObject,
                                             state.dataVariableSpeedCoils->VarSpeedCoil(DXCoilNum).Name));
-                    ShowContinueError(state, format("...{} output is not equal to 1.0 (+ or - 10%) at rated conditions.", cAlphaFields(10)));
+                    ShowContinueError(state, format("...{} output is not equal to 1.0 (+ or - 11%) at rated conditions.", cAlphaFields(11)));
```
and same here, 10%

rraustad reviewed (changes requested):
Wow, that's a lot of changes to add a curve object. There are 2 comments that should be addressed.

yujiex committed:
fix global replace err, change 11 back to 10

yujiex committed:
fix non-matching cAlphaFields and AlphArray

rraustad commented on src/EnergyPlus/VariableSpeedCoils.cc:
```diff
@@ -1182,7 +1182,7 @@ namespace VariableSpeedCoils {
                         state,
                         format("{}{}=\"{}\", invalid", RoutineName, CurrentModuleObject, state.dataVariableSpeedCoils->VarSpeedCoil(DXCoilNum).Name));
-                    ShowContinueError(state, format("...not found {}=\"{}\".", cAlphaFields(14), AlphArray(10)));
+                    ShowContinueError(state, format("...not found {}=\"{}\".", cAlphaFields(10), AlphArray(10)));
```
This looks correct now.

rraustad commented on src/EnergyPlus/VariableSpeedCoils.cc:
```diff
@@ -1717,7 +1717,7 @@ namespace VariableSpeedCoils {
                                                CurrentModuleObject,
                                                state.dataVariableSpeedCoils->VarSpeedCoil(DXCoilNum).Name));
-                        ShowContinueError(state, format("...not found {}=\"{}\".", cAlphaFields(AlfaFieldIncre), AlphArray(14 + (I - 1) * 6)));
+                        ShowContinueError(state, format("...not found {}=\"{}\".", cAlphaFields(AlfaFieldIncre), AlphArray(AlfaFieldIncre)));
```
This corrects a mistake...

rraustad commented on src/EnergyPlus/VariableSpeedCoils.cc:
```diff
@@ -1810,7 +1810,7 @@ namespace VariableSpeedCoils {
                                                CurrentModuleObject,
                                                state.dataVariableSpeedCoils->VarSpeedCoil(DXCoilNum).Name));
-                        ShowContinueError(state, format("...not found {}=\"{}\".", cAlphaFields(AlfaFieldIncre), AlphArray(16 + (I - 1) * 6)));
+                        ShowContinueError(state, format("...not found {}=\"{}\".", cAlphaFields(AlfaFieldIncre), AlphArray(AlfaFieldIncre)));
```
...and this corrects a mistake. These corrections should be described in the PR description. To be explicit, an issue would be created that this PR corrects.

yujiex replied:
Let me create an issue and link it to the PR

yujiex replied:
Created it: [#10137](https://github.com/NREL/EnergyPlus/issues/10137)

yujiex committed:
Merge remote-tracking branch 'origin/develop' into crankcaseHeater

yujiex committed:
min-field correction

yujiex committed:
fix unit test after merging

yujiex committed:
fix transition after merge conflict

yujiex committed:
Fix test idf after merge

yujiex commented:
Hi @rraustad and @Myoldmopar, I addressed the comments and merged in develop. Would you mind taking another look? Thanks

Myoldmopar commented:
@rraustad looks like you are satisfied based on the comments addressed above.  Could you mark it as approved if so?  I'll look over things and build/test now.

Myoldmopar commented on datasets/DXCoolingCoil.idf:
```diff
@@ -1358,6 +1363,7 @@
     DXCoilAirInletNode,      !- Air Inlet Node Name
     DXCoilAirOutletNode,     !- Air Outlet Node Name
     ,                        !- Crankcase Heater Capacity {W}
+    ,                        !- Crankcase Heater Capacity Function of Temperature Curve Name
```
OK, datasets updated manually with new field which will require transition.

rraustad reviewed (approved):
Comments have been addressed. This is ready to merge.

Myoldmopar commented on src/Transition/CreateNewIDFUsingRulesV23_2_0.f90:
```diff
@@ -493,6 +493,30 @@ SUBROUTINE CreateNewIDFUsingRules(EndOfFile,DiffOnly,InLfn,AskForInput,InputFile
                 OutArgs(23:CurArgs+4)=InArgs(19:CurArgs)
                 CurArgs = CurArgs + 4
 
+              CASE('COIL:COOLING:DX:CURVEFIT:PERFORMANCE')
```
I thought I was going to be merging this right now but this is another big set of transitions for the coils.  I need to look this over deeper and ... probably?  ... do some manual merging of these rules.

yujiex replied:
> This is still not an integer. Did you by chance forget to push a commit?

Let me see. I changed it in other places in the header but missed this line somehow.

Myoldmopar reviewed (commented):
I need to work on the transition stuff before I call it ready to go, but it certainly is close.

Myoldmopar commented:
> Comments have been addressed. This is ready to merge.

Thanks @rraustad, it's definitely close, but I need to spend some time with the transition code before merging.

yujiex committed:
crankcaseHeaterCapacityCurveIndex use int, not Real

Myoldmopar committed:
Merge remote-tracking branch 'origin/develop' into crankcaseHeater

Myoldmopar committed:
Blending transition rules together, cleaning up fortran, fixing bad versions in some example files

Myoldmopar commented:
OK, I spent a little time in there this evening.  As I did with the other coil transition branch, I ran transition on the full set of 23.1 example files, which passed fine.  I then ran EnergyPlus on the manually-converted example files in this branch, along with the set of auto-converted example files I had just made.  I compared the full set of ERR outputs and found no meaningful difference, meaning once again that the transition code appears to be working happily on an extensive array of configurations.  

While I was looking at the ERR file diff, I noticed some files were still 23.1 version, so I cleaned those up.  I also integrated the transition rule markdown file...the variable speed DX coil now has changes related to the 90.1 metrics, the latent/curve stuff, and now this crankcase change.  FYI @jmarrec !!!

As long as there isn't another coil transition PR in the queue, we should be good.  I am uninterested in doing this deep of a transition test again this week.

So I think everything is happy now.  I'll let CI run and check in the morning, but if it's all clean I plan to merge it in.

Myoldmopar commented:
And local regression testing is showing only 2 ERR diffs, which are introduced by me.  They are simply cleaning up the bad version number in a couple files.  Hopefully Decent agrees with me on that diff result.

Myoldmopar commented:
OK, so the failures showing up are also showing up on other branches.  Apparently a previous coil started producing duplicate output columns and mathdiff is fumbling.  Other than that, this looks clean though, so I'm merging it and I'll investigate that other issue.  Thanks @yujiex

yujiex commented:
Thanks @Myoldmopar  =)

jmarrec commented:
> While I was looking at the ERR file diff, I noticed some files were still 23.1 version, so I cleaned those up. I also integrated the transition rule markdown file...the variable speed DX coil now has changes related to the 90.1 metrics, the latent/curve stuff, and now this crankcase change. FYI @jmarrec !!!

@Myoldmopar Any specific action items for me to tackle here since you tagged me?
